### PR TITLE
Use Txid instead of UUID

### PR DESCRIPTION
**Pull request by gabre**
This PR changes the ID of a TX to be an SHA256 hash. We use the UTF8 byte form of the arguments from the Chaincode Invocation to generate the hash.
Note that there will be another PR changing Chaincode API to operate with only one byte array argument. This will make hashing easier (no need for UTF8 or any conversion). <PR_number_placeholder>
## Description

The following modifications are included:
- SHA256 hash for TX ID (not only the first 128 bits)
- Removal of previously used random generated UUID
- Renaming of UUID to TxID where it was needed
- It is still possible to add other custom TX ID generation algorithms (using TX payload) but this is the default.
## Motivation and Context

UUID cannot be calculated by client. This PR makes it possible:
- the user has to create byte arrays from the arguments of the chain code invocation
- the byte arrays has to be used to generate the SHA256 hash
## Fixes

No fixes, only new functionality.
## How Has This Been Tested?

Unit tests were run.
## Checklist:

- [x] I have added a [Signed-off-by](https://github.com/hyperledger/fabric/blob/master/CONTRIBUTING.md#legal-stuff).
- [x] I have either added documentation to cover my changes or this change requires no new documentation.
- [x] I have either added unit tests to cover my changes or this change requires no new tests.
- [x] I have run [golint](https://github.com/golang/lint) and have fixed valid warnings in code I have added or modified. This tool generates false positives so you may choose to ignore some warnings. The goal is clean, consistent, and readable code.

Signed-off-by:
Gabor Hosszu gabor@digitalasset.com

### CODE REVIEW

File: bddtests/syschaincode/noop/chaincode.go [hunk 1]
```diff
@@ -30,7 +30,7 @@ import (
 var logger = shim.NewLogger("noop")
 
 type ledgerHandler interface {
-	GetTransactionByUUID(txUUID string) (*protos.Transaction, error)
+	GetTransactionByID(txUUID string) (*protos.Transaction, error)
```
Comment (esgott) [inline bddtests/syschaincode/noop/chaincode.go]: I think the parameters should be renamed as well.

File: core/chaincode/handler.go [hunk 2]
```diff
@@ -331,7 +331,7 @@ func (handler *Handler) processStream() error {
 				chaincodeLogger.Debug("Next state nil message, ending chaincode support stream")
 				return err
 			}
-			chaincodeLogger.Debugf("[%s]Move state message %s", shortuuid(in.Uuid), in.Type.String())
+			chaincodeLogger.Debugf("[%s]Move state message %s", shorttxid(in.Uuid), in.Type.String())
```
Comment (esgott) [inline core/chaincode/handler.go]: why does in has Uuid?

File: core/chaincode/handler.go [hunk 3]
```diff
@@ -581,7 +581,7 @@ func (handler *Handler) afterGetState(e *fsm.Event, state string) {
 		e.Cancel(fmt.Errorf("Received unexpected message type"))
 		return
 	}
-	chaincodeLogger.Debugf("[%s]Received %s, invoking get state from ledger", shortuuid(msg.Uuid), pb.ChaincodeMessage_GET_STATE)
+	chaincodeLogger.Debugf("[%s]Received %s, invoking get state from ledger", shorttxid(msg.Uuid), pb.ChaincodeMessage_GET_STATE)
```
Comment (esgott) [inline core/chaincode/handler.go]: Why does msg has Uuid?

**Comment by gregmisiorek** [comment]
if all 3 checks have failed, but the branch has no conflict, then the fork is successful, isn't it?

**Comment by srderson** [comment]
@gabre @esgott Yes on those other UUID references. These point to https://github.com/hyperledger/fabric/blob/master/protos/chaincode.proto#L152 which likely should be changed to `txID` also. Confirmed with @muralisrini. Of course this is a breaking API change so that needs to be considered.

**Comment by srderson** [comment]
@gregmisiorek A couple of behave tests failed in Travis. Sometimes these flap due to timeouts so I will rerun ton confirm. Jenkins is not reliable yet.

**Comment by gabre** [comment]
@rameshthoomu You are right, those might be real failures. The PR changes the previous UUID mechanism to another which is called 'sha256base64' that assumes the 1st argument of the chaincode invocation to be a base64 encoded string. This should be changed/fixed somehow.

**Comment by gabre** [comment]
We could do the following changes to this PR:
- sha256 calculation: it can remain as it is (calculated from the 1st argument of the chaincode invocation... but this does not seem to be general) OR we can calculate that hash from e.g. the wire format version of the chaincode invocation (a byte array made by protobuf Marshal)
- base64: we should remove this. In case we use protobuf wire format for hash calculation, it is trivially removable.
  @tamasblummer @esgott What is your opinion?

File: bddtests/syschaincode/noop/chaincode.go (hunk 1 above)
Comment (gabre) [inline bddtests/syschaincode/noop/chaincode.go]: Done.

File: core/chaincode/handler.go (hunk 2 above)
Comment (gabre) [inline core/chaincode/handler.go]: Done.

File: core/chaincode/handler.go (hunk 3 above)
Comment (gabre) [inline core/chaincode/handler.go]: Done.

**Comment by gabre** [comment]
@srderson Could you check this please?

**Comment by gabre** [comment]
@srderson I rebased it.

**Comment by gabre** [comment]
'chaincode authorizable_counter with 4 peers, two stopped, bring back both' is this unstable? Before rebasing, it was OK.

**Comment by christo4ferris** [comment]
reverify Jenkins

**Comment by christo4ferris** [comment]
@gabre we will be transferring the fabric and fabric-api repository and all the corresponding issues and pull requests to the hyperledger-archives org on GH and replacing this repository with a read-only mirror of our Gerrit source. As a result, any automatic GH redirects will be broken. If you haven't already migrated your PR to Gerrit, and intend to do so, you will need to change your git remote to reference the transferred repository (e.g. git@github.com:hyperledger-archives/fabric) on your local clone.

**Pull request closed without merging**